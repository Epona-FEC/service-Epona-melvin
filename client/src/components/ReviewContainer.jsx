/* eslint-disable class-methods-use-this */
import React from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';

import sampleData from '../../sampleData';
import ImageCarousel from './ImageCarousel.jsx';
import ProductStats from './ProductStats.jsx';

import Review from './Review.jsx';

import '../style.css';

class ReviewContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      productReviews: {},
      seeMoreReviewsIsClicked: false,
      shopReviews: [],
      productReviewsSelected: true,
      reviewPhotos: [
        {
          url: 'https://source.unsplash.com/random/180x180',
          description: 'Purchase of a customer',
        },
        {
          url: 'https://source.unsplash.com/random/180x180',
          description: 'Purchase of a customer',
        },
        {
          url: 'https://source.unsplash.com/random/180x180',
          description: 'Purchase of a customer',
        },
        {
          url: 'https://source.unsplash.com/random/180x180',
          description: 'Purchase of a customer',
        },
        {
          url: 'https://source.unsplash.com/random/180x180',
          description: 'Purchase of a customer',
        },
      ],
      productReviewTotal: 25,
      shopReviewTotal: 50,
      shopAvgScore: 0,
      carouselTransform: 0,
      carouselWidth: null,
      loading: true,
    };
    this.getFirstFourReviews = this.getFirstFourReviews.bind(this);
    this.displayPrevImages = this.displayPrevImages.bind(this);
    this.displayNextImages = this.displayNextImages.bind(this);
    this.moreReviewsClick = this.moreReviewsClick.bind(this);
    this.getRestOfReviews = this.getRestOfReviews.bind(this);
  }

  componentDidMount() {
    const { reviewPhotos } = this.state;
    const { productId } = this.props;
    // Maxwidth is 810px and 16px margin
    const carouselWidth = Math.ceil(reviewPhotos.length / 4) * 840;
    this.setState({ carouselWidth });
    axios(`http://ec2-54-183-28-128.us-west-1.compute.amazonaws.com:3003/listing/${productId}`)
      .then(({ data }) => Promise.resolve(this.setState({
        productReviews: data.productReviews,
        shopReviews: data.shopReviews,
      })))
      .then(() => Promise.resolve(this.setState((state) => {
        const photos = [];
        state.productReviews.forEach((review) => {
          if (review.ReviewPhoto !== null) {
            photos.push({
              url: review.ReviewPhoto.url,
              description: review.ReviewPhoto.description,
            });
          }
        });
        let totalShopScore = 0;
        state.shopReviews.forEach((review) => {
          totalShopScore += review.score;
        });

        return {
          productReviewTotal: state.productReviews.length,
          shopReviewTotal: state.shopReviews.length,
          reviewPhotos: photos,
          shopAvgScore: totalShopScore / state.shopReviews.length,
        };
      })))
      .then(()=>this.setState({loading:false}))
      .catch((err) => {
        console.error(err);
      });
  }

  getFirstFourReviews(reviews) {
    const { renderStars } = this.props;
    const firstFour = [];
    for (let i = 0; i < 4; i += 1) {
      const reviewData = reviews[i];
      if (reviews[i] === undefined) {
        break;
      } else {
        console.log(reviewData);
        firstFour.push(<Review
          review={{body: reviewData.body, photoUrl: reviewData.ReviewPhoto.url, score: reviewData.score, date: reviewData.date }}
          product={reviewData.Product}
          reviewer={{
            avatar: reviewData.reviewer.photoUrl,
            username: reviewData.reviewer.name,
          }}
          renderStars={renderStars}
        />);
      }
    }
    return firstFour;
  }

  getRestOfReviews(reviews) {
    const { renderStars } = this.props;
    const rest = [];
    for (let i = 4; i < 20; i += 1) {
      const reviewData = reviews[i];
      if (reviews[i] === undefined) {
        break;
      } else {
        rest.push(<Review
          review={reviewData}
          product={reviewData.Product}
          reviewer={{
            avatar: reviewData.reviewer.photoUrl,
            username: reviewData.reviewer.name,
          }}
          renderStars={renderStars}
        />);
      }
    }
    return rest;
  }

  /**
   * Sets currrent tab, for reviews
   * @param {boolean} isSelected  Boolean if product reviews is selected
   *
   */
  setCurrentTab(isSelected) {
    this.setState({
      productReviewsSelected:
      isSelected,
    });
  }


  moreReviewsClick() {
    this.setState({ seeMoreReviewsIsClicked: true });
  }

  /**
   * Changes translateX value dynamically on carousel-slide class to the left
   */
  displayPrevImages() {
    const { carouselTransform } = this.state;
    if (carouselTransform === 0) {
      return;
    }
    this.setState((prevState) => ({
      carouselTransform: prevState.carouselTransform + 840,
      carouselWidth: prevState.carouselWidth + 840,
    }));
  }

  /**
   * Changes translateX value dynamically on carousel-slide class to the right
   */
  displayNextImages() {
    const { carouselWidth } = this.state;
    // Stops if reaches the end of carousel
    if (carouselWidth - 840 === 0) {
      return;
    }
    this.setState((prevState) => ({
      carouselTransform: prevState.carouselTransform - 840,
      carouselWidth: prevState.carouselWidth - 840,
    }));
  }

  render() {
    const { renderStars } = this.props;
    const {
      productReviewTotal, shopReviewTotal, loading,
      productReviews, shopReviews, productReviewsSelected,
      seeMoreReviewsIsClicked, reviewPhotos, carouselTransform, shopAvgScore,
    } = this.state;

    return (loading)? <div></div>:(
      <div className='review-container'>
        <ProductStats renderStars={renderStars} totalShopReviews={shopReviewTotal} reviewScore={ shopAvgScore } />
        <div className="review-tabs">
          <button
            type="button"
            className={`review-button ${productReviewsSelected ? 'active-button' : 'non-active-button'}`}
            onClick={() => this.setState({ productReviewsSelected: true })}
          >
            Reviews for this item
            <span className='review-data'>
              {productReviewTotal}
            </span>
          </button>
          <button
            type="button"
            className={`review-button ${!productReviewsSelected ? 'active-button' : 'non-active-button'}`}
            onClick={() => this.setState({ productReviewsSelected: false })}
          >
            Reviews for this shop
            <span className='review-data'>{shopReviewTotal}</span>
          </button>
        </div>
        {/* If product reviews selected render product reviews */}
        {(productReviewsSelected) ? [
          <Reviews getReviews={this.getFirstFourReviews} reviews={productReviews} />,
          (productReviews.length > 4 && !seeMoreReviewsIsClicked)
            ? <div className='see-more-container'><button className="more-reviews-button" type="button" onClick={() => this.moreReviewsClick()}>See More Reviews</button></div> : null,
          (seeMoreReviewsIsClicked)
            ? <Reviews getReviews={this.getRestOfReviews} reviews={productReviews} /> : null,
        ] : [
        // If shopReviews is selected:
          // Render Reviews component with shopReviews
          <Reviews getReviews={this.getFirstFourReviews} reviews={shopReviews} />,
          (shopReviews.length > 4 && !seeMoreReviewsIsClicked) ? <div className='see-more-container'><button className="more-reviews-button" type="button" onClick={this.moreReviewsClick}>See More Reviews</button></div> : null,
          (seeMoreReviewsIsClicked)
            ? <Reviews getReviews={this.getRestOfReviews} reviews={shopReviews} /> : null,
        ]}
        {/* Picture Carousel here */}
        <ImageCarousel
          photos={reviewPhotos}
          styles={{ transform: `translateX(${carouselTransform}px)` }}
          slideInPrevImages={this.displayPrevImages}
          slideInNextImages={this.displayNextImages}
        />
      </div>
    );
  }
}
// TODO: Create component for this Reviews Component
const Reviews = ({ getReviews, reviews }) => getReviews(reviews).map((data) => data);

ReviewContainer.propTypes = {
  renderStars: PropTypes.func,
  productId: PropTypes.number,
};

ReviewContainer.defaultProps = {
  renderStars: () => null,
  productId: 5,
};

export default ReviewContainer;
