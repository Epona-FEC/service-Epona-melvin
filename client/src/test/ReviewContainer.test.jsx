/* eslint-disable no-undef */
import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import ReviewContainer from '../components/ReviewContainer.jsx';

Enzyme.configure({ adapter: new Adapter() });

describe('ReviewContainer', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ReviewContainer />);
  });

  it('Should shallow, intial sampledata', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('See more button is visible if more than 4 reviews', () => {
    expect(wrapper.contains(<button className="more-reviews-button" type="button">See More</button>)).toBe(true);
  });

  // Test see more button not to be there when state array is less than 4
  it('See more button is visible if more than 4 reviews', () => {
    wrapper.setState({productReviews: [1, 2, 3] });
    expect(wrapper.contains(<button className="more-reviews-button" type="button">See More</button>)).toBe(false);
  });

  // Reviews for items should be selected intially
  it('Initial state for selected reviews should be product', () => {
    expect(wrapper.state().selectedReviews).toBe('product');
  });

  // When Reviews for shop is clicked, selected reviews state should change
  it('When Reviews for shop is clicked, selected reviews should be shop', () => {
    wrapper.find('.shop-review-button').simulate('click');
    expect(wrapper.state().selectedReviews).toBe('shop');
  });

  // https://www.etsy.com/listing/737189527/crystal-bridal-headband-crystal-bridal?ga_order=most_relevant&ga_search_type=all&ga_view_type=gallery&ga_search_query=&ref=sc_gallery-1-1&plkey=1d3e3fe8f64cca5ea1d347e588198c8c1651a52e%3A737189527&pro=1&frs=1

  // Has more than four reviews and less than 20, shows the rest of the array when see more is clicked

  //https://www.etsy.com/listing/535133523/new-colors-unicorn-horns-indvidual-or?ga_order=most_relevant&ga_search_type=all&ga_view_type=gallery&ga_search_query=&ref=sr_gallery-1-5&bes=1&col=1

  // Has more than 20 reviews, when see more is clicked a Read All reviews is shown

});
