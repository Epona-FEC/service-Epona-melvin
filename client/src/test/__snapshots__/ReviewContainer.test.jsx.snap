// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewContainer Should shallow, intial sampledata 1`] = `
<div
  className="review-container"
>
  <div
    className="review-tabs"
  >
    <button
      className="product-review-button"
      onClick={[Function]}
      type="button"
    >
      Reviews for this item
      <span>
         
        25
      </span>
    </button>
    <button
      className="shop-review-button"
      onClick={[Function]}
      type="button"
    >
      Reviews for this shop
      <span>
        50
      </span>
    </button>
  </div>
  <Review
    product={
      Object {
        "id": 1,
        "name": "Product #1",
        "photoUrl": "https://loremflickr.com/50/50",
      }
    }
    review={
      Object {
        "body": "This a test",
        "date": 2020-03-17T23:08:19.000Z,
        "photoUrl": "https://source.unsplash.com/random/200x200",
        "score": 5,
      }
    }
    reviewer={
      Object {
        "avatar": "https://i.pravatar.cc/50",
        "username": "Albert",
      }
    }
  />
  <Review
    product={
      Object {
        "id": 2,
        "name": "Product #2",
        "photoUrl": "https://loremflickr.com/50/50",
      }
    }
    review={
      Object {
        "body": "This a test",
        "date": 2020-03-17T23:08:19.000Z,
        "photoUrl": "https://source.unsplash.com/random/200x200",
        "score": 5,
      }
    }
    reviewer={
      Object {
        "avatar": "https://i.pravatar.cc/50",
        "username": "Joe",
      }
    }
  />
  <Review
    product={
      Object {
        "id": 3,
        "name": "Product #3",
        "photoUrl": "https://loremflickr.com/50/50",
      }
    }
    review={
      Object {
        "body": "This a test",
        "date": 2020-03-17T23:08:19.000Z,
        "photoUrl": "https://source.unsplash.com/random/200x200",
        "score": 5,
      }
    }
    reviewer={
      Object {
        "avatar": "https://i.pravatar.cc/50",
        "username": "Ro",
      }
    }
  />
  <Review
    product={
      Object {
        "id": 3,
        "name": "Product #4",
        "photoUrl": "https://loremflickr.com/50/50",
      }
    }
    review={
      Object {
        "body": "This a test",
        "date": 2020-03-17T23:08:19.000Z,
        "photoUrl": "https://source.unsplash.com/random/200x200",
        "score": 5,
      }
    }
    reviewer={
      Object {
        "avatar": "https://i.pravatar.cc/50",
        "username": "Po",
      }
    }
  />
  <button
    className="more-reviews-button"
    type="button"
  >
    See More
  </button>
</div>
`;
